### PR TITLE
[docs] Improve Batch Training on Core example

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

## Why are these changes needed?

Applies feedback from a late review - should make the example simpler and easier to follow. Includes some restructuring, addition of an overview at the top, improvements to poor naming in code.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 1]
```diff
@@ -254,125 +292,70 @@
...
     "    transform_time = time.time() - start_time\n",
-    "    print_time(f\"Data transform time for LocID: {i}: {transform_time}\")\n",
+    "    print_time(f\"Data transform time for LocID: {location_id}: {transform_time}\")\n",
     "\n",
-    "    return file_name, i, train_and_evaluate(data, models, i)"
+    "    return (\n",
+    "        file_name,\n",
+    "        location_id,\n",
+    "        train_and_evaluate_internal(data, models, location_id),\n",
+    "    )"
```
Comment (matthewdeng) [inline doc/source/ray-core/examples/batch_training.ipynb:326]: For consistency with the above steps, how about extracting the timing steps into this function?

Something like:
```python
start_time = time.time()
results = train_and_evaluate_internal(data, models, location_id)
train_and_evaluate_time = time.time() - start_time
print_time(...)

return (file_name, location_id, results)
```

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 2]
```diff
@@ -383,7 +366,9 @@
     "    # Sleep a moment for nicer output\n",
     "    time.sleep(1)\n",
     "    print(\"\", flush=True)\n",
-    "    print(f\"Total number of models (all tasks): {count_not_none} ({count})\")\n",
+    "    print(\n",
+    "        f\"Total number of models (all tasks): {count_not_none * len(models)} ({count})\"\n",
+    "    )\n",
```
Comment (matthewdeng) [inline doc/source/ray-core/examples/batch_training.ipynb:371]: Nit: It's not really obvious what's going on here when it's actually output.

Example from below:
```
Total number of models (all tasks): 1566 (522)
```

I think it would be simpler to just print everything out separately, e.g:
```
Number of models train per pickup location:
Number of pickup locations: 
Number of pickup locations with enough records to train:
```

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 3]
```diff
@@ -397,7 +382,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 12,
+   "execution_count": 10,
```
Comment (matthewdeng) [inline doc/source/ray-core/examples/batch_training.ipynb:385]: For the text in the section above, can you introduce that this is the part where we start stitching all the code together to actually run the job?

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 4]
```diff
@@ -31,6 +31,13 @@
...
+    "# Contents\n",
+    "In this this tutorial, you will learn about:\n",
+    " 1. Reading parquet data,\n",
+    " 2. Creating Ray tasks to preprocess, train and evaluate data batches,\n",
+    " 3. Dividing data into batches and spawning a Ray task for each batch to be ran in parallel,\n",
+    " 4. The runtime-memory tradeoff between delegating data loading to each task as opposed to centralized data loading.\n",
+    "\n",
     "# Walkthrough\n",
     "\n",
     "Our task is to create separate time series models for each pickup location. We can use the `pickup_location_id` column in the dataset to group the dataset into data batches. We will then fit models for each batch and choose the best one.\n",
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply doc/source/ray-core/examples/batch_training.ipynb:43]: Also worth mentioning that this is split by month?

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 5]
```diff
@@ -459,27 +447,35 @@
...
      ]
     }
    ],
    "source": [
     "from sklearn.tree import DecisionTreeRegressor\n",
     "\n",
-    "results = run(\n",
+    "results = run_batch_training(\n",
```
Comment (matthewdeng) [inline doc/source/ray-core/examples/batch_training.ipynb:477]: Should we include a section describing what to do with the results of this training job?

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 6]
```diff
@@ -557,13 +591,30 @@
     "    )\n",
     "    load_tasks = [read_into_object_store_pg.remote(file) for file in files]\n",
```
Comment (matthewdeng) [inline doc/source/ray-core/examples/batch_training.ipynb:592]: You should be able to just run the tasks with `scheduling_strategy="SPREAD"` and not set a placement group.

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 7]
```diff
@@ -557,13 +591,30 @@
...
     "    group_refs = {}\n",
-    "    for i, refs in enumerate(ray.get(load_tasks)):\n",
-    "        group_refs[files[i]] = refs\n",
+    "    for location_id, refs in enumerate(ray.get(load_tasks)):\n",
```
Comment (matthewdeng) [inline doc/source/ray-core/examples/batch_training.ipynb:594]: This should be `file_id` instead of `location_id`. Though in general it'd be helpful to give all the `refs` more descriptive names so it's clear what is the object(s) being referenced.

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 8]
```diff
@@ -493,11 +489,11 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## Loading data once into Ray object store\n",
+    "## Centralized data loading using Ray object store\n",
```
Comment (matthewdeng) [inline doc/source/ray-core/examples/batch_training.ipynb:492]: Prefix this section with something along the lines of `[Optional]` or `[Optimization]` ?

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 9]
```diff
@@ -31,6 +31,13 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
+    "# Contents\n",
+    "In this this tutorial, you will learn about:\n",
+    " 1. Reading parquet data,\n",
+    " 2. Creating Ray tasks to preprocess, train and evaluate data batches,\n",
+    " 3. Dividing data into batches and spawning a Ray task for each batch to be ran in parallel,\n",
+    " 4. The runtime-memory tradeoff between delegating data loading to each task as opposed to centralized data loading.\n",
```
Comment (matthewdeng) [inline doc/source/ray-core/examples/batch_training.ipynb:39]: Can we add something like these as headers within the Walkthrough so that it's easier for the reader to understand what each step is in the walkthrough?

**Comment by Yard1** [comment]
@matthewdeng Applied feedback, PTAL!

**Review by richardliaw — APPROVED** [review]
Overall looks great! 

Can you make sure you keep a 80char line width? It's hard to parse otherwise when rendered on docs.

**Review by matthewdeng — APPROVED** [review]
Looks great, few nits. Thanks for making the improvements!

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 10]
```diff
@@ -219,25 +262,27 @@
...
-    "    df: pd.DataFrame, models: List[BaseEstimator], i: int = 0\n",
+    "def train_and_evaluate_internal(\n",
+    "    df: pd.DataFrame, models: List[BaseEstimator], pickup_location_id: int = 0\n",
     ") -> List[Tuple[BaseEstimator, float]]:\n",
     "    # We need at least 4 rows to create a train / test split.\n",
     "    if len(df) < 4:\n",
-    "        print_time(f\"Dataframe for LocID: {i} is empty or smaller than 4\")\n",
+    "        print(\n",
+    "            f\"Dataframe for LocID: {pickup_location_id} is empty or smaller than 4\"\n",
+    "        )\n",
```
[the remark this answers is not in the document]
Comment (Yard1) [reply doc/source/ray-core/examples/batch_training.ipynb:283]: I don't think we can do relative references inside code cells.

[the remark this answers is not in the document]
Comment (jjyao) [reply doc/source/ray-core/examples/batch_training.ipynb:283]: I feel this is generally an issue, I'm also wondering whether our linkcheck checks links in comments. @maxpumperla what's the best practices here?

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 11]
```diff
@@ -493,23 +517,43 @@
...
     "\n",
-    "Notice we do not call `ray.get()` on the references of the `read_into_object_store`. Instead, we pass the reference itself as the argument to the `task.remote` dispatch, [allowing for the data to stay in the object store until it is actually needed](ray-pass-large-arg-by-value). This avoids a situation where all the data would be loaded into the memory of the process calling `ray.get()`.\n",
+    "Notice we do not call `ray.get()` on the references of the `read_into_object_store`. Instead, we pass the reference itself as the argument to the `train_and_evaluate.remote` dispatch, [allowing for the data to stay in the object store until it is actually needed](ray-pass-large-arg-by-value). This avoids a situation where all the data would be loaded into the memory of the process calling `ray.get()`.\n",
```
Comment (jjyao) [inline doc/source/ray-core/examples/batch_training.ipynb:524]: I think you are talking about a different pattern: https://docs.ray.io/en/master/ray-core/patterns/unnecessary-ray-get.html

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 12]
```diff
@@ -529,66 +573,111 @@
...
-    "    for i in loc_list:\n",
+    "    data_batch_refs = []\n",
+    "    for pickup_location_id in pickup_location_ids:\n",
     "        # Put each data batch as a separate dataframe into Ray object store.\n",
-    "        group_refs.append(\n",
-    "            (i, ray.put(locdf.filter(pc.field(\"pickup_location_id\") == i).to_pandas()))\n",
+    "        data_batch_refs.append(\n",
+    "            (\n",
+    "                pickup_location_id,\n",
+    "                ray.put(\n",
```
Comment (jjyao) [inline doc/source/ray-core/examples/batch_training.ipynb:584]: @stephanie-wang do you think we should use the experimental dynamic returns (https://docs.ray.io/en/master/ray-core/tasks/generators.html) in the public examples?

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 13]
```diff
@@ -529,66 +573,111 @@
...
      ]
     }
    ],
    "source": [
-    "results = run(files, task_generator_with_object_store, models=[LinearRegression()])\n",
+    "results = run_batch_training_with_object_store(\n",
+    "    files, models=[LinearRegression()]\n",
+    ")\n",
```
Comment (jjyao) [inline doc/source/ray-core/examples/batch_training.ipynb:680]: For line 688, technically we have object spilling for datasets that cannot fit in the memory so it's possible that large dataset it's still faster to load once. I think it's more practical to say that people can try both way and pick the approach that's best for their workload.

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 14]
```diff
@@ -529,66 +573,111 @@
...
-    "    pg = placement_group([{\"CPU\": 1}] * len(files), strategy=\"SPREAD\")\n",
-    "    ray.get(pg.ready())\n",
+    "    read_into_object_store_spread = read_into_object_store.options(\n",
+    "        scheduling_strategy=\"SPREAD\"\n",
+    "    )\n",
+    "    read_tasks = [read_into_object_store_spread.remote(file) for file in files]\n",
+    "\n",
+    "    # Dictionary of references to data batches with file names as keys\n",
+    "    data_batch_refs_by_file = {}\n",
+    "    for file_id, data_batch_refs in enumerate(ray.get(read_tasks)):\n",
```
Comment (jjyao) [inline doc/source/ray-core/examples/batch_training.ipynb:616]: Not sure if we want to make it more complicated, but there is another optimization we can do here: we don't need to wait until all read_tasks to finish before submitting `train_and_evaluate`, as long as some read_task is finished, we can submit `train_and_evaluate`: https://docs.ray.io/en/master/ray-core/patterns/ray-get-submission-order.html

Comment (Yard1) [reply doc/source/ray-core/examples/batch_training.ipynb:616]: I think it would make it too complex, yeah

File: doc/source/ray-core/examples/batch_training.ipynb [hunk 15]
```diff
@@ -31,9 +31,17 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
+    "# Contents\n",
+    "In this this tutorial, we will explore:\n",
```
Comment (richardliaw) [inline doc/source/ray-core/examples/batch_training.ipynb:35]:
```suggestion
    "In this tutorial, we will walk through the following steps:\n",
```

File: doc/source/ray-core/examples/batch_training.ipynb (hunk 12 above)
Comment (stephanie-wang) [reply doc/source/ray-core/examples/batch_training.ipynb:584]: Hmm yeah it'd be good to use it if we can. Or if the num_returns can be determined dynamically, we can just use the `return list` syntax.
